src/thor/costmatrix.cc

```diff
@@ -948,82 +956,93 @@ void CostMatrix::SetTargets(baldr::GraphReader& graphreader,
-      const DirectedEdge* opp_edge = nullptr;
-      GraphId opp_edge_id = graphreader.GetOpposingEdgeId(edgelabel.edgeid(), opp_edge, tile);
-      if (opp_edge == nullptr) {
-        throw tile_gone_error_t("CostMatrix::RecostPaths failed", edgelabel.edgeid());
-      }
+    if (edge->is_shortcut()) {
+      auto superseded = graphreader.RecoverShortcut(edgelabel.edgeid());
+      recovered_inner_edges.insert(superseded.begin() + 1, superseded.end());
+      std::move(superseded.rbegin(), superseded.rend(), std::back_inserter(path_edges));
+    } else
+      path_edges.push_back(edgelabel.edgeid());
+  }
 
-      if (opp_edge->is_shortcut()) {
-        auto superseded = graphreader.RecoverShortcut(opp_edge_id);
-        recovered_inner_edges.insert(superseded.begin() + 1, superseded.end());
-        std::move(superseded.begin(), superseded.end(), std::back_inserter(path_edges));
-      } else
-        path_edges.emplace_back(std::move(opp_edge_id));
+  // Reverse the list
+  std::reverse(path_edges.begin(), path_edges.end());
+
+  // Append the reverse path from the destination - use opposing edges
+  // The first edge on the reverse path is the same as the last on the forward
+  // path, so get the predecessor.
+  auto& target_edgelabels = edgelabel_[MATRIX_REV][target_idx];
+  for (auto edgelabel_index = target_edgelabels[connedge_idx2].predecessor();
+       edgelabel_index != kInvalidLabel;
+       edgelabel_index = target_edgelabels[edgelabel_index].predecessor()) {
+    const BDEdgeLabel& edgelabel = target_edgelabels[edgelabel_index];
+    const DirectedEdge* opp_edge = nullptr;
+    GraphId opp_edge_id = graphreader.GetOpposingEdgeId(edgelabel.edgeid(), opp_edge, tile);
+    if (opp_edge == nullptr) {
+      throw tile_gone_error_t("CostMatrix::RecostPaths failed", edgelabel.edgeid());
     }
 
+    if (opp_edge->is_shortcut()) {
+      auto superseded = graphreader.RecoverShortcut(opp_edge_id);
+      recovered_inner_edges.insert(superseded.begin() + 1, superseded.end());
+      std::move(superseded.begin(), superseded.end(), std::back_inserter(path_edges));
+    } else
+      path_edges.emplace_back(std::move(opp_edge_id));
+  }
+
+  auto source_edge = find_correlated_edge(source, path_edges.front());
+  auto target_edge = find_correlated_edge(target, path_edges.back());
+  float source_pct = static_cast<float>(source_edge.percent_along());
+  float target_pct = static_cast<float>(target_edge.percent_along());
+
+  // TODO(nils): bug with trivial routes https://github.com/valhalla/valhalla/issues/4433
+  // remove this whole block below once that's fixed
+  if (path_edges.size() == 1 && source_pct > target_pct) {
+    // it found the wrong direction, so let's turn that around
+    auto opp_id = graphreader.GetOpposingEdgeId(path_edges[0]);
+    path_edges.clear();
+    path_edges.emplace_back(opp_id);
+    source_pct = 1.f - source_pct;
+    target_pct = 1.f - target_pct;
+  }
```

Comment:
this is quite messed up, see https://github.com/valhalla/valhalla/issues/4433. I think it'd be easy-ish to fix before here and I'll do that. but this works quite well, even though you'll see in the test that there's some tiny floating point issue apparently